File: scray-querying/modules/scray-sync-api/src/main/scala/scray/sync/api/VersionedData.scala
/**
// see the licence.txt file distributed with this work for additional
// information regarding copyright ownership.
//
// licensed under the apache license, version 2.0 (the "license");
// you may not use this file except in compliance with the license.
// you may obtain a copy of the license at
// http://www.apache.org/licenses/license-2.0
// unless required by applicable law or agreed to in writing, software
// distributed under the license is distributed on an "as is" basis,
// without warranties or conditions of any kind, either express or implied.
// see the license for the specific language governing permissions and
// limitations under the license.

// see the licence.txt file distributed with this work for additional
// information regarding copyright ownership.
//
// licensed under the apache license, version 2.0 (the "license");
// you may not use this file except in compliance with the license.
// you may obtain a copy of the license at

// http://www.apache.org/licenses/license-2.0

// unless required by applicable law or agreed to in writing, software
// distributed under the license is distributed on an "as is" basis,
// without warranties or conditions of any kind, either express or implied.
// see the license for the specific language governing permissions and
// limitations under the license.

package scray.sync.api

class versioneddata(
  val datasource: string,  // defines where the data come from. e.g. from a batch or streaming job
  val mergekey: string,    // describes a attribute to merge two datasources
  val version: long,       // version of this data. e.g. time stamp
  val data: string         // string representation of the data to store
) {

  def getdataas[t](f: string => t): t = {
    f(data)
  }

  def getversionkey: int = {
    versioneddata.createversionkey(datasource, mergekey)
  }

  def getdatasource: string = {
    datasource
  }

  def getmergekey: string = {
    mergekey
  }

  def getversion: long = {
    version
  }

  def getdata: string = {
    data
  }

  override def tostring: string = {
    s"""{"datasource: "${datasource}", "mergekey": "${mergekey}", "version": ${version}, "data": "${data}"}"""
  }
}

object versioneddata {
  def createversionkey(datasource: string, mergekey: string): int = {
    datasource.hashcode() * 31 + mergekey.hashcode() * 31
  }
}
**/

File: scray-querying/modules/scray-sync-api/src/main/scala/scray/sync/api/VersionedData.java
// See the LICENCE.txt file distributed with this work for additional
// information regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// See the LICENCE.txt file distributed with this work for additional
// information regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

// http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package scray.sync.api;

import java.util.function.Function;

public class VersionedData {
    private final String dataSource;
    private final String mergeKey;
    private final long version;
    private final String data;

    public VersionedData(String dataSource, String mergeKey, long version, String data) {
        this.dataSource = dataSource;
        this.mergeKey = mergeKey;
        this.version = version;
        this.data = data;
    }

    public <T> T getDataAs(Function<String, T> f) {
        return f.apply(data);
    }

    public int getVersionKey() {
        return VersionedData.createVersionKey(dataSource, mergeKey);
    }

    public String getDataSource() {
        return dataSource;
    }

    public String getMergeKey() {
        return mergeKey;
    }

    public long getVersion() {
        return version;
    }

    public String getData() {
        return data;
    }

    @Override
    public String toString() {
        return String.format("{\"dataSource\": \"%s\", \"mergeKey\": \"%s\", \"version\": %d, \"data\": \"%s\"}",
                dataSource, mergeKey, version, data);
    }

    public static int createVersionKey(String dataSource, String mergeKey) {
        return dataSource.hashCode() * 31 + mergeKey.hashCode() * 31;
    }
}